pages/api/submit.ts

```diff
@@ -110,24 +120,123 @@ const previewNotify = async (form: PublicFormSchemaProperties, fields: Responses
+        if (arrayValue.length === 0 || typeof arrayValue[0] === "string") {
+          return {
+            ...previousValueResolved,
+            fields: {
+              ...previousValueResolved.fields,
+              [current]: arrayValue as string[],
+            },
+          };
+        }
+        // otherwise we assume its a file
+        else if (typeof arrayValue[0] === "object") {
+          return {
+            ...previousValueResolved,
+            files: {
+              ...previousValueResolved.files,
+              [current]: await Promise.all(
+                arrayValue.map(async (fileObj) => {
+                  return await processFileData(fileObj as FileInputResponse);
+                })
+              ),
+            },
+          };
+        }
+      } else if (typeof keyPairValue === "object") {
+        return {
+          ...previousValueResolved,
+          files: {
+            ...previousValueResolved.files,
+            [current]: await processFileData(keyPairValue as FileInputResponse),
+          },
+        };
+      }
+      return previousValueResolved;
+    },
+    Promise.resolve({
+      fields: {},
+      files: {},
+    } as SubmissionParsedRequest)
+  );
+};
+
+/**
+ * This function will take a FileInputResponse object and return File object should
+ * the file conform to expected mimetypes and size. Otherwise an error will be thrown
+ * @param fileObj
+ */
+const processFileData = async (fileObj: FileInputResponse): Promise<ProcessedFile> => {
+  // if we have a size key present in the data then we can simply throw an error if
+  // that number is greater than 8mb. We do not depend on this however. This is simply
+  // to be more efficient. Regardless if this parameter is less than 8mb the size will still be checked
+  if (typeof fileObj?.size === "number" && fileObj.size / 1048576 > 8) {
+    throw new Error("FileSizeError: A file has been uploaded that is greater than 8mb in size");
+  }
+  // process Base64 encoded data
+  if (typeof fileObj?.name === "string" && typeof fileObj?.file === "string") {
+    // base64 string should be data URL https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs
+    const fileData = fileObj.file.match(/^data:([A-Za-z-+/]+);base64,(.+)$/);
+    if (fileData?.length !== 3) {
+      throw new Error(`FileTypeError: The file ${fileObj.name} was not a valid data URL`);
     }
+    // create buffer with fileData
+    const fileBuff = Buffer.from(fileData[2], "base64");
+    // if the buffer is larger then 8mb then this is larger than the filesize that's allowed
+    if (Buffer.byteLength(fileBuff) / 1048576 > 8) {
+      throw new Error(
+        `FileSizeError: The file ${fileObj.name} has been uploaded that is greater than 8mb in size`
+      );
+    }
+    // determine the real mime type of the file from the buffer
+    const mimeType = await fileTypeFromBuffer(fileBuff);
```

Comment:
The problem here is that the API is open. Theoretically you can submit to the API completely outside of the client. For client side it's okay to just check the provided file type because we're using the `accepts` parameter on the input. For the API we cannot trust for certain that the file type provided is the actual file type for the file and so we need to infer it to ensure it's actually the case. 

Good catch on the text-based format files. I tested with a pdf and it seemed to work file although this is not text based. I will test with those file formats

Reply:
In terms of the strategy I meant they should be the same, not necessarily to trust the client side.  So if we're inferring server-side we should also be using the same strategy of inferring client side.